mono/utils/mono-threads-api.h

```diff
@@ -140,6 +140,22 @@ For further explanation of what can and can't be done in GC unsafe mode:
 		mono_threads_exit_gc_safe_region_unbalanced_internal (__gc_safe_unbalanced_cookie, &__gc_safe_unbalanced_dummy);	\
 	} while (0)
 
+void
+mono_threads_enter_no_safepoints_region (const char *func);
+
+void
+mono_threads_exit_no_safepoints_region (const char *func);
+
+#define MONO_ENTER_NO_SAFEPOINTS					\
+	do {								\
+		if (mono_threads_are_safepoints_enabled ())		\
+			mono_threads_enter_no_safepoints_region (__func__)
+
+#define MONO_EXIT_NO_SAFEPOINTS							\
+		if (mono_threads_are_safepoints_enabled ())			\
+			mono_threads_exit_no_safepoints_region (__func__);	\
```

Comment:
I wonder if this even worth a macro.
I get that the braces force a balance.
But one could just call mono_threads_enter/exit_no_safepoints_region () and avoid the ugly.
In C++ the end macro would just a brace -- destructor.

Reply:
there are a couple of reasons for the macro:
1. I want to block scope these things.
2. if we decide we want to allow nesting no-safepoints scopes, in C, I'm going to need to declare a var to propagate some state from the enter to the exit.
3. consistency with the other `MONO_ENTER_*`/`MONO_EXIT_*` macros.

Reply:
All three reasons are really: C needs them, C++ does not.